Runs tests on the repository.

jfboismenu opened the pull request:
This adds support for running tests on multiple platforms and with different Python interpreters. It also pushes code coverage to `codecov`. Note that Azure's Linux with Python 2.7 is currently skipped due to Python being compiled with USC4 characters and the pip installable versions of PySide are USC2 compatible only.

The documentation is also build and we're running the pre-commit hooks on the repo instead of running black and flake8 manually, which makes more sense as each repo has it's own exclusion list and passing the parameters to the build system feels clunky. It's much better to simple use the exclusion list already provided inside the pre-commit configuration file.

As you can clearly see, debugging CI is quite exhausting and time consuming.

Commit:
code style validation refactored into it's own file.

Commit:
fixed include

Commit:
fixed structure in template

Commit:
triggering failure

Commit:
trying out the exclude_flake8 parameter. Forgot to test it.

Commit:
adding extra param

Commit:
passing in via template substitution

Commit:
flake8 errors should reappear now

Commit:
added run_tests, doesn't do much for now

Commit:
fixing pip install, probably will work now?

Commit:
upgrading pip

Commit:
using python -m syntax

Commit:
--user on pip install

jfboismenu committed:
Update unit_tests.yml

jfboismenu committed:
Update unit_tests.yml

jfboismenu committed:
Update unit_tests.yml

Commit:
trying out the fixed pip install from tk-toolchain

Commit:
restructuring the pip installs

Commit:
switching to PySide1 for now, as PySide2 binairies for Python 2.7
are not pip installable it seems.

Commit:
reinstating PySide2, it should work, because it worked before.

Commit:
something is up with python 2 and the use python task, it seems
forcing the Python version prevents PySide2 from installing?

Commit:
trying task again.

Commit:
removing python task switcher. it seems Python 2.7.16 can't pip install
PySide2.

Commit:
putting back python task selector

Commit:
upgrading pip

Commit:
wip

Commit:
removing python -m

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
trying Qt's wheels

Commit:
Checking out 3.7 build

Commit:
tk-toolchain should install everything we need

Commit:
trying 2.7 again.

Commit:
trying something else

Commit:
wip

Commit:
trimming strategy

Commit:
one last try

Commit:
wip

Commit:
wip

Commit:
templating python version

Commit:
template paths are relative apparently

Commit:
fixing python versions

Commit:
using template notation to expand parameters.

Commit:
adding parameters.<>

Commit:
forgot () around qtLibs

Commit:
bashening

Commit:
overhaul of how PySide is getting install

Commit:
fixing post install on Linux?

Commit:
removing python in front of post install script for PySide

Commit:
import PySide instead for the test

Commit:
wip

Commit:
switching ubuntu

Commit:
one last try, for real now.

Commit:
almost works

Commit:
refactoring a bit

Commit:
adding gui test

Commit:
adding parameters.

Commit:
adding Qt.py for tests

Commit:
fixing parameter expansion

Commit:
parameters passing doesn't seem to work...

Commit:
forgot to pass parameters through

Commit:
adding macos and Linux

Commit:
launching xvfb and restructured the pip installs

Commit:
environment->env

Commit:
reactivating other builds and adding comments

Commit:
uploading coverage to azure

Commit:
using a version of core where gitattributes keep the correct formatting
for pickle files

Commit:
verbose output for PyTest

Commit:
adding coveralls

Commit:
fixing syntax

Commit:
passing in true string.

Commit:
always pass in template parameters!

Commit:
debugging upload_coverage

Commit:
fixed syntax

Commit:
fixed shell command

Commit:
stringifying true

Commit:
removing tempalte stuff

Commit:
one more try

Commit:
wip

Commit:
coverage outside unit tests yml

Commit:
wip

Commit:
moving back inside the file

Commit:
maybe?

Commit:
refactored into full templates

Commit:
trying something new

Commit:
fixing typo

Commit:
fixing indent

Commit:
passing in job name

Commit:
renamed all templates

Commit:
updated doc

Commit:
updated doc

Commit:
will it work?

Commit:
fixed indentation

Commit:
removing coveralls

Commit:
added some comments

Commit:
some polish

Commit:
more comments

Commit:
switches back to master for tk-toolchain

Commit:
triggered!

jfboismenu commented on templates/code-style-validation.yml:
```diff
@@ -0,0 +1,51 @@
+
+  # Ensure the code that was pushed has the proper formatting, except in the
+  # ui folder, which is auto-generated.
+  - bash: black --check . --exclude="ui\/.*py$"
```
I think at some point it might be better to `pip install pre-commit` and run `pre-commit run -all` on the repo instead, which would catch all other error types. Plus, each repo will already have their own exclusion list properly set up in the pre-commit configuration file, so we won't have to try and be clever here.

Commit:
polish comments

Commit:
done?

Commit:
updated default azure-pipelines.yml snippet

Commit:
removed flake8 from build

Commit:
adding documentation validation

Commit:
adds documenation test

jfboismenu commented on build-pipeline.yml:
```diff
@@ -9,41 +9,13 @@
-  # Install all the Python code validation packages.
-  - bash: python -m pip install black flake8
-    displayName: Install validators
-
```
this got refactored into `code-style-validation.yml`. Note that the refactored version just runs the pre-commit hook now. If the user was properly set up, then the pre-commit checks should pass.

jfboismenu commented on templates/code-style-validation.yml:
```diff
@@ -0,0 +1,31 @@
+      pre-commit run --all
+    displayName: Validate code
+
+  - bash: (test -e docs && tk-docs-preview --build-only) || echo No "docs" folder detected.
```
I also added documentation generation. We will never be caught unaware that the documentation does not build just before release now. 😄

Commit:
forgot to pip install tk-toolchain to generate doc

willis102 commented on README.md:
```diff
@@ -2,9 +2,9 @@
+# How to use the build system.
 
-Create a file named  `azure-pipeline.yml` file at the root of your repo and put the following:
+Create a file named `azure-pipelines.yml` file at the root of your repo and put the following in it:
```
Should we also mention creating / copying `.pre-commit-config.yaml` in the README?  Now that code-style-validation relies on the pre-commit hook, this seems like something worth explaining in the documentation.

jfboismenu replied:
Done

willis102 reviewed (approved):
Sorry for the delay on this one!  This stuff is awesome JF.

One question below.

Commit:
fixed doc generation

Commit:
added section about pre-commit

Commit:
wip of toolchain fix

Commit:
fixed build policy

Commit:
forgot to install tk-core for doc generation

Commit:
need to clone tk-core from master, obviously...

Commit:
can now parametrize where tk-toolchain is coming from

Commit:
testing the tk_toolchain_ref argument

Commit:
forgot a &&

Commit:
removed test value. it worked!
[skip ci]

Commit:
renamed extra_pip_dependencies to extra_test_dependencies

jfboismenu commented:
Since the last code review, I've added a new parameter to build-pipelines, named `tk_toolchain_ref`, which allows to override which tk-toolchain branch to use. This will avoid having to create a branch for `tk-ci-tools` just to test a bugfix inside tk-toolchain. I've also renamed `extra_pip_dependencies` to `extra_test_dependencies` for more clarity.

The build fails right now because the doc validation step is clone tk-toolchain from master, which is missing a bugfix. I've tested the fix using `tk_toolchain_ref` and it worked however.

willis102 commented on azure-pipelines.yml:
```diff
@@ -8,14 +8,21 @@
   tags:
     include:
     - v*
+pr:
```
👍

willis102 commented on templates/code-style-validation.yml:
```diff
@@ -28,6 +32,8 @@ jobs:
-      ((test -e docs && tk-docs-preview --build-only) || echo No "docs" folder detected.)
+      # PySide 2 and tk-toolchain are required to build the documentation.
+      pip install PySide2 git+https://github.com/shotgunsoftware/tk-toolchain.git@${{ parameters.tk_toolchain_ref }}#egg=tk-toolchain &&
+      git clone https://github.com/shotgunsoftware/tk-core --depth 1 --branch SG-13264-python-3-compat ../tk-core &&
```
Does this warrant a TODO or a comment about why we're using a branch of tk-core for now?

jfboismenu replied:
It should, but I already have another branch ready for review that addresses this question 😄

willis102 replied:
I can't keep up!

willis102 reviewed (approved):
LGTM!  The documentation changes are awesome!

willis102 commented on README.md:
```diff
@@ -1,12 +1,76 @@
+Tools for Azure Pipelines. These tools allow a Toolkit developer to validate their code using Azure Pipelines, ensuring the code is formatted properly, does not contain ambiguous statements and passes all the tests.
 
-# How to use these tools
+The pipeline is split into two main sections
```
❤️ The docs you added are great

Commit:
trigger CI
